### PR TITLE
Add __declspec(dllexport) to class Node and class Op if build on Wind…

### PR DESCRIPTION
**Pull request by HectorSVC**
Add __declspec(dllexport) to class Node and class Op if build on Windows, so that it can be referenced in some test code in MXNet

### CODE REVIEW

**Comment by HectorSVC** [comment]
I got error while building MXNet with test projects on Windows. I figured out there's some changes needed in NNVM code as well as MXNet code itself.
5>     Creating library D:/Git/MXNet/build/tests/Debug/mxnet_unit_tests.lib and object D:/Git/MXNet/build/tests/Debug/mxnet_unit_tests.exp
5>coreop_perf.obj : error LNK2019: unresolved external symbol "public: static class nnvm::Op const * __cdecl nnvm::Op::Get(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?Get@Op@nnvm@@SAPEBV12@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "public: void __cdecl mxnet::test::op::CoreOpExecutor<float>::Init(class std::vector<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >,class std::allocator<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > > > const &,class std::vector<class mxnet::NDArray,class std::allocator<class mxnet::NDArray> > const &,class std::vector<class mxnet::NDArray,class std::allocator<class mxnet::NDArray> > const &,class mxnet::test::op::CoreOpExecutor<float> const *)" (?Init@?$CoreOpExecutor@M@op@test@mxnet@@QEAAXAEBV?$vector@U?$pair@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V12@@std@@V?$allocator@U?$pair@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V12@@std@@@2@@std@@AEBV?$vector@VNDArray@mxnet@@V?$allocator@VNDArray@mxnet@@@std@@@6@1PEBV1234@@Z)
5>core_op_runner_test.obj : error LNK2001: unresolved external symbol "public: static class nnvm::Op const * __cdecl nnvm::Op::Get(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?Get@Op@nnvm@@SAPEBV12@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
5>operator_tune_test.obj : error LNK2001: unresolved external symbol "public: static class nnvm::Op const * __cdecl nnvm::Op::Get(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?Get@Op@nnvm@@SAPEBV12@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
5>coreop_perf.obj : error LNK2019: unresolved external symbol "private: static class dmlc::any const * __cdecl nnvm::Op::GetAttrMap(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?GetAttrMap@Op@nnvm@@CAPEBVany@dmlc@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "public: static class nnvm::OpMap<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > const & __cdecl nnvm::Op::GetAttr<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??$GetAttr@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Op@nnvm@@SAAEBV?$OpMap@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@1@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
5>core_op_runner_test.obj : error LNK2001: unresolved external symbol "private: static class dmlc::any const * __cdecl nnvm::Op::GetAttrMap(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?GetAttrMap@Op@nnvm@@CAPEBVany@dmlc@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
5>operator_tune_test.obj : error LNK2001: unresolved external symbol "private: static class dmlc::any const * __cdecl nnvm::Op::GetAttrMap(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?GetAttrMap@Op@nnvm@@CAPEBVany@dmlc@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
5>coreop_perf.obj : error LNK2019: unresolved external symbol "private: static void __cdecl nnvm::Op::UpdateAttrMap(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::function<void __cdecl(class dmlc::any *)>)" (?UpdateAttrMap@Op@nnvm@@CAXAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V?$function@$$A6AXPEAVany@dmlc@@@Z@4@@Z) referenced in function "public: static class nnvm::OpMap<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > const & __cdecl nnvm::Op::GetAttr<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??$GetAttr@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Op@nnvm@@SAAEBV?$OpMap@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@1@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
5>core_op_runner_test.obj : error LNK2001: unresolved external symbol "private: static void __cdecl nnvm::Op::UpdateAttrMap(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::function<void __cdecl(class dmlc::any *)>)" (?UpdateAttrMap@Op@nnvm@@CAXAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V?$function@$$A6AXPEAVany@dmlc@@@Z@4@@Z)
5>operator_tune_test.obj : error LNK2001: unresolved external symbol "private: static void __cdecl nnvm::Op::UpdateAttrMap(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::function<void __cdecl(class dmlc::any *)>)" (?UpdateAttrMap@Op@nnvm@@CAXAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V?$function@$$A6AXPEAVany@dmlc@@@Z@4@@Z)
5>coreop_perf.obj : error LNK2019: unresolved external symbol "public: static class std::shared_ptr<class nnvm::Node> __cdecl nnvm::Node::Create(void)" (?Create@Node@nnvm@@SA?AV?$shared_ptr@VNode@nnvm@@@std@@XZ) referenced in function "private: class std::shared_ptr<class nnvm::Node> __cdecl mxnet::test::op::CoreOpExecutor<float>::MakeNode(void)const " (?MakeNode@?$CoreOpExecutor@M@op@test@mxnet@@AEBA?AV?$shared_ptr@VNode@nnvm@@@std@@XZ)
5>core_op_runner_test.obj : error LNK2001: unresolved external symbol "public: static class std::shared_ptr<class nnvm::Node> __cdecl nnvm::Node::Create(void)" (?Create@Node@nnvm@@SA?AV?$shared_ptr@VNode@nnvm@@@std@@XZ)
5>operator_tune_test.obj : error LNK2001: unresolved external symbol "public: static class std::shared_ptr<class nnvm::Node> __cdecl nnvm::Node::Create(void)" (?Create@Node@nnvm@@SA?AV?$shared_ptr@VNode@nnvm@@@std@@XZ)
5>D:\Git\MXNet\build\tests\Debug\mxnet_unit_tests.exe : fatal error LNK1120: 4 unresolved externals

**Comment by HectorSVC** [comment]
it's different, we should not add extern "C". 
sample in MXNet here:
https://github.com/apache/incubator-mxnet/search?utf8=%E2%9C%93&q=MXNET_API&type=

or we should change 
#define NNVM_DLL NNVM_EXTERN_C __declspec(dllexport)
to 
#define NNVM_DLL __declspec(dllexport) //remove NNVM_EXTERN_C

**Comment by HectorSVC** [comment]
@tqchen please review again

File: include/nnvm/c_api.h [hunk 1]
```diff
@@ -39,7 +39,7 @@ typedef void *GraphHandle;
  * \brief Set the last error message needed by C API
  * \param msg The error message to set.
  */
-NNVM_DLL void NNAPISetLastError(const char* msg);
+NNVM_EXTERN_C NNVM_DLL void NNAPISetLastError(const char* msg);
```
[the remark this answers is not in the document]
Comment (tqchen) [reply include/nnvm/c_api.h]: remove define of NNVM_EXTERN_C, instead use
```c++
#ifdef __cplusplus
extern "C" {
#endif

NNVM_DLL func1();
...
#ifdef __cplusplus
}
#endif

```

[the remark this answers is not in the document]
Comment (HectorSVC) [reply include/nnvm/c_api.h]: yes, that's the correct way, the existing code of this part is wired.

**Comment by HectorSVC** [comment]
seems the build hangs on deploy.

**Comment by HectorSVC** [comment]
@tqchen any future comments?

**Pull request merged**